src/Common/Services/EnvironmentDeploymentHandler.cs

```diff
@@ -187,22 +195,46 @@ public async Task Handle(bool interactive, CustomerEnvironment customerEnvironme
                 var result = await startDeploymentInput.StartDeploymentAsync(true);
             }
 
-            // show progress from deployment
-            TimeSpan timeout = TimeSpan.FromHours(1);
-            using (CancellationTokenSource cancellationTokenSource = new CancellationTokenSource(timeout))
+            
+            using (CancellationTokenSource cancellationTokenMainTask = new CancellationTokenSource(timeoutMainTask))
             {
+                // The variable 'utcOfLastMessageReceived' will be set with the UTC of last message received on message bus (on ProcessDeploymentMessage()).
+                // This 'cancellationTokenMBMessageReceived' will be restarted if the time past between 'utcOfLastMessageReceived' and the current datetime
+                // is less than 'timeoutToGetSomeMBMessageTask'.
+                CancellationTokenSource cancellationTokenMBMessageReceived = new CancellationTokenSource(timeoutToGetSomeMBMessageTask);
+
                 while (!this._isDeploymentFinished)
                 {
                     _session.LogPendingMessages();
+
                     try
                     {
-                        await Task.Delay(TimeSpan.FromSeconds(0.1), cancellationTokenSource.Token);
+                        await Task.Delay(TimeSpan.FromSeconds(0.1), cancellationTokenMBMessageReceived.Token);
                     }
                     catch (TaskCanceledException)
                     {
-                        break;
+
+                        if (cancellationTokenMBMessageReceived.Token.IsCancellationRequested)
+                        {
+                            if (utcOfLastMessageReceived != null && (DateTime.UtcNow - utcOfLastMessageReceived < timeoutToGetSomeMBMessageTask))
+                            {
+                                cancellationTokenMBMessageReceived = new CancellationTokenSource(timeoutToGetSomeMBMessageTask);
+                            }

```

Comment:
This still doesn't shorten the time at the end of it, you need to remove the time spent waiting from the new cancelation token.
```suggestion
                            if (utcOfLastMessageReceived != null)
                            {
                                TimeSpan timeWaited = DateTime.UtcNow - utcOfLastMessageReceived.Value;
                                if (timeWaited < timeoutToGetSomeMBMessageTask)
                                {
                                    cancellationTokenMBMessageReceived = new CancellationTokenSource(timeoutToGetSomeMBMessageTask - timeWaited);
                                }
                            }

```

Reply:
Since I don't know the source of the request, can't really tell what you are trying to achieve.
But if you want to have a CancellationToken that can depend on more than one factor you can try to use the CancellationTokenSource.

A good example here:
https://dotnetcodr.com/2014/05/02/cancelling-a-task-with-a-composite-cancellation-token-in-net-c/

Reply:
I will add your change locally and test. But yeah, makes sense to only wait 15 min without receive any message.

Reply:
@jpsfs Thx for your feedback, can be interesting to apply and useful!